In [ ]:
!pip install ydata-profiling
!pip install --upgrade Pillow

  Using cached ydata_profiling-4.9.0-py2.py3-none-any.whl (356 kB)
  Using cached pydantic-2.8.2-py3-none-any.whl (423 kB)
  Using cached dacite-1.8.1-py3-none-any.whl (14 kB)
  Using cached htmlmin-0.1.12-py3-none-any.whl
  Using cached typeguard-4.3.0-py3-none-any.whl (35 kB)
  Using cached wordcloud-1.9.3-cp39-cp39-win_amd64.whl (300 kB)
  Using cached ImageHash-4.3.1-py2.py3-none-any.whl (296 kB)
  Using cached phik-0.12.4-cp39-cp39-win_amd64.whl (666 kB)
  Using cached numba-0.60.0-cp39-cp39-win_amd64.whl (2.7 MB)
  Using cached llvmlite-0.43.0-cp39-cp39-win_amd64.whl (28.1 MB)
  Using cached annotated_types-0.7.0-py3-none-any.whl (13 kB)
  Attempting uninstall: llvmlite
    Found existing installation: llvmlite 0.38.0


ERROR: Cannot uninstall 'llvmlite'. It is a distutils installed project and thus we cannot accurately determine which files belong to it which would lead to only a partial uninstall.


In [ ]:
import numpy as np
import pandas as pd
from ydata_profiling import ProfileReport
import os as os

ModuleNotFoundError: No module named 'ydata_profiling'

Read dataset

In [ ]:
df = pd.read_csv("CARES_data_cleaned.csv")
df

Indexno   Age  Gender  RCRI score Anemia category  PreopEGFRMDRD  \
0            2  48.0  Female         NaN             NaN            NaN   
1            5  36.0  Female         NaN             NaN            NaN   
2            6  64.0  Female         NaN            Mild     152.538570   
3            9  73.0    Male         NaN        Moderate     117.231496   
4           10  73.0    Male         0.0            Mild      98.651255   
...        ...   ...     ...         ...             ...            ...   
90780   121192  66.0  Female         2.0            Mild      48.339582   
90781   121194  50.0    Male         1.0        Moderate     126.592489   
90782   121197  58.0  Female         NaN             NaN      86.306771   
90783   121198  63.0  Female         0.0             NaN            NaN   
90784   121200  45.0  Female         0.0            Mild     125.902498   

      GradeofKidneydisease  DaysbetweenDeathandoperation  @30daymortality  \
0                      NaN                           NaN            False   
1                      NaN                           NaN            False   
2                       G1                           NaN            False   
3                       G1                           NaN            False   
4                       G1                          59.0            False   
...                    ...                           ...              ...   
90780                  G3a                           NaN            False   
90781                   G1                           NaN            False   
90782                   G2                           NaN            False   
90783                  NaN                           NaN            False   
90784                   G1                           NaN            False   

       Preoptransfusionwithin30days  ...  \
0                               0.0  ...   
1                               0.0  ...   
2                               0.0  ...   
3                               0.0  ...   
4                               0.0  ...   
...                             ...  ...   
90780                           0.0  ...   
90781                           0.0  ...   
90782                           0.0  ...   
90783                           0.0  ...   
90784                           0.0  ...   

       Anemiacategorybinned_Moderate/Severe  Anemiacategorybinned_NA  \
0                                     False                     True   
1                                     False                     True   
2                                     False                    False   
3                                      True                    False   
4                                     False                    False   
...                                     ...                      ...   
90780                                 False                    False   
90781                                  True                    False   
90782                                 False                     True   
90783                                 False                     True   
90784                                 False                    False   

       RDW15.7_<= 15.7 RDW15.7_>15.7 RDW15.7_NA ASAcategorybinned_I  \
0                False         False       True                True   
1                 True         False      False                True   
2                 True         False      False                True   
3                 True         False      False                True   
4                False          True      False               False   
...                ...           ...        ...                 ...   
90780             True         False      False               False   
90781             True         False      False               False   
90782             True         False      False               False   
90783            False         False       True               False   
90784 

In [ ]:
df.describe()

Indexno           Age    RCRI score  PreopEGFRMDRD  \
count   90785.000000  90785.000000  63361.000000   79955.000000   
mean    60334.001057     52.253225      0.322296      96.426155   
std     34930.467992     17.087307      0.634789      33.954241   
min         2.000000     18.000000      0.000000       2.541026   
25%     29973.000000     39.000000      0.000000      79.078715   
50%     60572.000000     54.000000      0.000000      96.398561   
75%     90788.000000     65.000000      1.000000     114.310169   
max    121200.000000    103.000000      6.000000     671.298147   

       DaysbetweenDeathandoperation  Preoptransfusionwithin30days  \
count                   5595.000000                  90785.000000   
mean                     476.782127                      0.042672   
std                      421.866740                      0.405840   
min                        0.000000                      0.000000   
25%                      121.000000                      0.000000   
50%                      355.000000                      0.000000   
75%                      746.000000                      0.000000   
max                     1783.000000                     21.000000   

            Intraop  Postopwithin30days  Transfusionintraandpostop  
count  90785.000000        90785.000000               90785.000000  
mean       0.056276            0.018946                   0.075222  
std        0.230455            0.260728                   0.395623  
min        0.000000            0.000000                   0.000000  
25%        0.000000            0.000000                   0.000000  
50%        0.000000            0.000000                   0.000000  
75%        0.000000            0.000000                   0.000000  
max        1.000000           23.000000                  24.000000

In [ ]:
# Check number of missing values in each column
missing_values = df.isna().sum()
print(missing_values)

Indexno                        0
Age                            0
Gender                         0
RCRI score                 27424
Anemia category            66916
                           ...  
ASAcategorybinned_I            0
ASAcategorybinned_II           0
ASAcategorybinned_III          0
ASAcategorybinned_IV-VI        0
ASAcategorybinned_NA           0
Length: 94, dtype: int64


In [ ]:
# Convert the result to a DataFrame for better visualization
missing_values_df = missing_values.reset_index()
missing_values_df.columns = ['Column', 'Missing Values']

display(missing_values_df)

Column  Missing Values
0                   Indexno               0
1                       Age               0
2                    Gender               0
3                RCRI score           27424
4           Anemia category           66916
..                      ...             ...
89      ASAcategorybinned_I               0
90     ASAcategorybinned_II               0
91    ASAcategorybinned_III               0
92  ASAcategorybinned_IV-VI               0
93     ASAcategorybinned_NA               0

[94 rows x 2 columns]

In [ ]:
# Convert the result to a DataFrame for better visualization
missing_values_df = missing_values.reset_index()
missing_values_df.columns = ['Column', 'Missing Values']

# Export to Excel
missing_values_df.to_excel('missing_values.xlsx', index=False)

Remove columns that have been one-hot encoded, to prevent multicollinearity issues

In [ ]:
df.drop(['Gender', 'Anemia category', 'GradeofKidneydisease', 'AnaestypeCategory', 'PriorityCategory',
        'TransfusionIntraandpostopCategory', 'AGEcategory', 'AGEcategoryOriginal', 'SurgRiskCategory',
       'RaceCategory', 'GradeofKidneyCategory', 'Anemiacategorybinned', 'RDW15.7', 'ASAcategorybinned'],
       axis=1, inplace = True)

In [ ]:
df.head()

Indexno   Age  RCRI score  PreopEGFRMDRD  DaysbetweenDeathandoperation  \
0        2  48.0         NaN            NaN                           NaN   
1        5  36.0         NaN            NaN                           NaN   
2        6  64.0         NaN     152.538570                           NaN   
3        9  73.0         NaN     117.231496                           NaN   
4       10  73.0         0.0      98.651255                          59.0   

   @30daymortality  Preoptransfusionwithin30days  Intraop  Postopwithin30days  \
0            False                           0.0      0.0                 0.0   
1            False                           0.0      0.0                 0.0   
2            False                           0.0      0.0                 0.0   
3            False                           0.0      1.0                 0.0   
4            False                           0.0      0.0                 0.0   

   Transfusionintraandpostop  ...  Anemiacategorybinned_Moderate/Severe  \
0                        0.0  ...                                 False   
1                        0.0  ...                                 False   
2                        0.0  ...                                 False   
3                        1.0  ...                                  True   
4                        0.0  ...                                 False   

   Anemiacategorybinned_NA RDW15.7_<= 15.7 RDW15.7_>15.7 RDW15.7_NA  \
0                     True           False         False       True   
1                     True            True         False      False   
2                    False            True         False      False   
3                    False            True         False      False   
4                    False           False          True      False   

  ASAcategorybinned_I ASAcategorybinned_II  ASAcategorybinned_III  \
0                True                False                  False   
1                True                False                  False   
2                True                False                  False   
3                True                False                  False   
4               False                 True                  False   

   ASAcategorybinned_IV-VI  ASAcategorybinned_NA  
0                    False                 False  
1                    False                 False  
2                    False                 False  
3                    False                 False  
4                    False                 False  

[5 rows x 80 columns]

Mean imputation of missing values (for numerical columns **RCRI score** and **PreopEGFRMDRD**)

In [ ]:
# calculate the mean of RCRI score column
mean_RCRI = df['RCRI score'].mean()

mean_RCRI

0.32229604962042896

In [ ]:
df['RCRI score'].fillna(mean_RCRI, inplace = True)

In [ ]:
df['RCRI score']

0        0.322296
1        0.322296
2        0.322296
3        0.322296
4        0.000000
           ...   
90780    2.000000
90781    1.000000
90782    0.322296
90783    0.000000
90784    0.000000
Name: RCRI score, Length: 90785, dtype: float64

In [ ]:
# calculate the mean of PreopEGFRMDRD column

mean_PreopEGFRMDRD = df['PreopEGFRMDRD'].mean()
mean_PreopEGFRMDRD

96.42615515896141

In [ ]:
df['PreopEGFRMDRD'].fillna(mean_PreopEGFRMDRD, inplace = True)

In [ ]:
df['PreopEGFRMDRD']

0         96.426155
1         96.426155
2        152.538570
3        117.231496
4         98.651255
            ...    
90780     48.339582
90781    126.592489
90782     86.306771
90783     96.426155
90784    125.902498
Name: PreopEGFRMDRD, Length: 90785, dtype: float64

In [ ]:
df.describe()

Indexno           Age    RCRI score  PreopEGFRMDRD  \
count   90785.000000  90785.000000  90785.000000   90785.000000   
mean    60334.001057     52.253225      0.322296      96.426155   
std     34930.467992     17.087307      0.530313      31.864673   
min         2.000000     18.000000      0.000000       2.541026   
25%     29973.000000     39.000000      0.000000      81.773853   
50%     60572.000000     54.000000      0.000000      96.426155   
75%     90788.000000     65.000000      0.322296     111.439680   
max    121200.000000    103.000000      6.000000     671.298147   

       DaysbetweenDeathandoperation  Preoptransfusionwithin30days  \
count                   5595.000000                  90785.000000   
mean                     476.782127                      0.042672   
std                      421.866740                      0.405840   
min                        0.000000                      0.000000   
25%                      121.000000                      0.000000   
50%                      355.000000                      0.000000   
75%                      746.000000                      0.000000   
max                     1783.000000                     21.000000   

            Intraop  Postopwithin30days  Transfusionintraandpostop  
count  90785.000000        90785.000000               90785.000000  
mean       0.056276            0.018946                   0.075222  
std        0.230455            0.260728                   0.395623  
min        0.000000            0.000000                   0.000000  
25%        0.000000            0.000000                   0.000000  
50%        0.000000            0.000000                   0.000000  
75%        0.000000            0.000000                   0.000000  
max        1.000000           23.000000                  24.000000

In [ ]:
#Check that there are no more missing values in these two columns:
print(df['RCRI score'].isna().sum())
print(df['PreopEGFRMDRD'].isna().sum())


0
0


Median imputation of missing values (for numerical columns **RCRI score** and **PreopEGFRMDRD**)

In [ ]:
# Read original dataset first
df = pd.read_csv("CARES_data_cleaned.csv")
df

Indexno   Age  Gender  RCRI score Anemia category  PreopEGFRMDRD  \
0            2  48.0  Female         NaN             NaN            NaN   
1            5  36.0  Female         NaN             NaN            NaN   
2            6  64.0  Female         NaN            Mild     152.538570   
3            9  73.0    Male         NaN        Moderate     117.231496   
4           10  73.0    Male         0.0            Mild      98.651255   
...        ...   ...     ...         ...             ...            ...   
90780   121192  66.0  Female         2.0            Mild      48.339582   
90781   121194  50.0    Male         1.0        Moderate     126.592489   
90782   121197  58.0  Female         NaN             NaN      86.306771   
90783   121198  63.0  Female         0.0             NaN            NaN   
90784   121200  45.0  Female         0.0            Mild     125.902498   

      GradeofKidneydisease  DaysbetweenDeathandoperation  @30daymortality  \
0                      NaN                           NaN            False   
1                      NaN                           NaN            False   
2                       G1                           NaN            False   
3                       G1                           NaN            False   
4                       G1                          59.0            False   
...                    ...                           ...              ...   
90780                  G3a                           NaN            False   
90781                   G1                           NaN            False   
90782                   G2                           NaN            False   
90783                  NaN                           NaN            False   
90784                   G1                           NaN            False   

       Preoptransfusionwithin30days  ...  \
0                               0.0  ...   
1                               0.0  ...   
2                               0.0  ...   
3                               0.0  ...   
4                               0.0  ...   
...                             ...  ...   
90780                           0.0  ...   
90781                           0.0  ...   
90782                           0.0  ...   
90783                           0.0  ...   
90784                           0.0  ...   

       Anemiacategorybinned_Moderate/Severe  Anemiacategorybinned_NA  \
0                                     False                     True   
1                                     False                     True   
2                                     False                    False   
3                                      True                    False   
4                                     False                    False   
...                                     ...                      ...   
90780                                 False                    False   
90781                                  True                    False   
90782                                 False                     True   
90783                                 False                     True   
90784                                 False                    False   

       RDW15.7_<= 15.7 RDW15.7_>15.7 RDW15.7_NA ASAcategorybinned_I  \
0                False         False       True                True   
1                 True         False      False                True   
2                 True         False      False                True   
3                 True         False      False                True   
4                False          True      False               False   
...                ...           ...        ...                 ...   
90780             True         False      False               False   
90781             True         False      False               False   
90782             True         False      False               False   
90783            False         False       True               False   
90784 

In [ ]:
# calculate the median of RCRI score column, and replace missing values
median_RCRI = df['RCRI score'].median()
df['RCRI score'].fillna(median_RCRI, inplace = True)
df['RCRI score']

0        0.0
1        0.0
2        0.0
3        0.0
4        0.0
        ... 
90780    2.0
90781    1.0
90782    0.0
90783    0.0
90784    0.0
Name: RCRI score, Length: 90785, dtype: float64

In [ ]:
# calculate the median of PreopEGFRMDRD column, and replace missing values
median_PreopEGFRMDRD = df['PreopEGFRMDRD'].median()
df['PreopEGFRMDRD'].fillna(median_PreopEGFRMDRD, inplace = True)
df['PreopEGFRMDRD']

0         96.398561
1         96.398561
2        152.538570
3        117.231496
4         98.651255
            ...    
90780     48.339582
90781    126.592489
90782     86.306771
90783     96.398561
90784    125.902498
Name: PreopEGFRMDRD, Length: 90785, dtype: float64

In [ ]:
print("Mean RCRI score = " + str(mean_RCRI))
print("Median RCRI score = " + str(median_RCRI))
print("Mean PreopEGFRMDRD = " + str(mean_PreopEGFRMDRD))
print("Median PreopEGFRMDRD = " + str(median_PreopEGFRMDRD))

Mean RCRI score = 0.32229604962042896
Median RCRI score = 0.0
Mean PreopEGFRMDRD = 96.42615515896141
Median PreopEGFRMDRD = 96.39856108377884


In [ ]:
df.describe()

Indexno           Age    RCRI score  PreopEGFRMDRD  \
count   90785.000000  90785.000000  90785.000000   90785.000000   
mean    60334.001057     52.253225      0.224938      96.422863   
std     34930.467992     17.087307      0.550574      31.864675   
min         2.000000     18.000000      0.000000       2.541026   
25%     29973.000000     39.000000      0.000000      81.773853   
50%     60572.000000     54.000000      0.000000      96.398561   
75%     90788.000000     65.000000      0.000000     111.439680   
max    121200.000000    103.000000      6.000000     671.298147   

       DaysbetweenDeathandoperation  Preoptransfusionwithin30days  \
count                   5595.000000                  90785.000000   
mean                     476.782127                      0.042672   
std                      421.866740                      0.405840   
min                        0.000000                      0.000000   
25%                      121.000000                      0.000000   
50%                      355.000000                      0.000000   
75%                      746.000000                      0.000000   
max                     1783.000000                     21.000000   

            Intraop  Postopwithin30days  Transfusionintraandpostop  
count  90785.000000        90785.000000               90785.000000  
mean       0.056276            0.018946                   0.075222  
std        0.230455            0.260728                   0.395623  
min        0.000000            0.000000                   0.000000  
25%        0.000000            0.000000                   0.000000  
50%        0.000000            0.000000                   0.000000  
75%        0.000000            0.000000                   0.000000  
max        1.000000           23.000000                  24.000000

The mean of the **RCRI score** decreased (from 0.322 to 0.225) after replacing missing values in the **RCRI score** column with the median.